### DISCUSSION TITLE
problems with -resize not working

### DISCUSSION THREAD

**Discussion by flyerx** [ImageMagick/ImageMagick#3494] [Help] [answered]
Hello,

ImageMagick-7.0.11-5-Q16-HDRI-x64 (portable and Win64 dynamic install package)
Windows 10 (build 19042.906)

Using powershell and command prompt (in admin and regular modes). 

Using the portable version I tried to use the resize command after a couple of successful commands on the console as follows:

`"C:\Progs\ImageMagick-7.0.11-5-portable-Q16-HDRI-x64\magick" logo: logo.png`
`"C:\Progs\ImageMagick-7.0.11-5-portable-Q16-HDRI-x64\magick" logo.png   logo.jpg`
`"C:\Progs\ImageMagick-7.0.11-5-portable-Q16-HDRI-x64\magick" logo.jpg -resize 100x100 result.jpg`

the first two lines work properly but the 3rd line fails and does not produce an image or an error. The commands were ran on the root of drive E where I put all my work files. Putting the full path on the images does not change the results. Putting quotes around the 100x100 does not affect the issue.   After digging through the windows event logger magick.exe showed the following event:

>Faulting application name: magick.exe, version: 0.0.0.0, time stamp: 0x6060aa72
Faulting module name: magick.exe, version: 0.0.0.0, time stamp: 0x6060aa72
Exception code: 0xc0000409
Fault offset: 0x000000000032f569
Faulting process id: 0x3284
Faulting application start time: 0x01d72742cabd2e7b
Faulting application path: C:\Progs\ImageMagick-7.0.11-5-portable-Q16-HDRI-x64\magick.exe
Faulting module path: C:\Progs\ImageMagick-7.0.11-5-portable-Q16-HDRI-x64\magick.exe
Report Id: 58c6aa47-8233-4f00-9273-f65cccf3a6bc
Faulting package full name: 
Faulting package-relative application ID: 

I tried the same commands with the install package (ImageMagick-7.0.11-5-Q16-HDRI-x64-dll.exe) and the behavior was the same. The event logger showed this:

>Faulting application name: magick.exe, version: 0.0.0.0, time stamp: 0x6060a889
Faulting module name: ucrtbase.dll, version: 10.0.19041.789, time stamp: 0x2bd748bf
Exception code: 0xc0000409
Fault offset: 0x000000000007286e
Faulting process id: 0x321c
Faulting application start time: 0x01d7273fa931e12d
Faulting application path: C:\Program Files\ImageMagick-7.0.11-Q16-HDRI\magick.exe
Faulting module path: C:\WINDOWS\System32\ucrtbase.dll
Report Id: 14968e7e-95f4-43ae-b6cb-2222f68ca38a
Faulting package full name: 
Faulting package-relative application ID: 

I updated to the latest VC redistributable for 64 bit systems as recommended in the install page. I also checked the windows files using the sfc command just in case there were corrupted libraries.

I did notice that the -scale command worked but this may be a sign that I may be using the program wrong or that there is a problem with my installation. 

Any insight on this issue would be appreciated. 

Thanks

**Comment by snibgo** [comment]
Try running this:
```
set MAGICK_OCL_DEVICE=off
```
... and then the resizing command. Does that fix the problem?

**Reply by flyerx** [reply]
Same problem. The event logger came back with the same Exception code and offset.

**Comment by snibgo** [comment]
Please paste the full text output from "magick -version".

**Reply by flyerx** [reply]
Version: ImageMagick 7.0.11-5 Q16 x64 2021-03-28 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Visual C++: 192829912
Features: Cipher DPC HDRI OpenCL
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp xml zip zlib

**Comment by snibgo** [comment, marked as answer]
Okay, that confirms your build has OpenCL, which I thought might be the problem. But the first test would seem to eliminate that.

Try:
```
set MAGICK_OCL_DEVICE=off

"C:\Progs\ImageMagick-7.0.11-5-portable-Q16-HDRI-x64\magick" logo.jpg -resize 100x100 x1.jpg

"C:\Progs\ImageMagick-7.0.11-5-portable-Q16-HDRI-x64\magick" logo.jpg -blur 0x3 x2.jpg

"C:\Progs\ImageMagick-7.0.11-5-portable-Q16-HDRI-x64\magick" logo.jpg -evaluate Divide 2 x3.jpg
```
Do any of those magick commands work? Do any fail?

**Comment by flyerx** [comment]
Thanks. 
They work now. So what does the MAGICK_OCL_DEVICE=off do?

**Comment by snibgo** [comment]
Bingo! That command tells ImageMagick not to use OpenCL. OpenCL uses the GPU (Graphics Processing Unit) for some operations, like "-resize" and "-blur". This can be much faster than using the CPU. Sadly, it doesn't always work. When it doesn't work, IM should realise, and fall-back to the CPU, but this also doesn't always work.

You can set this as a system-wide environment variable. For Windows 8.1, one way to do that is Control panel => System => Advanced system settings => Advanced tab => Environment variables. Add a variable called "MAGICK_OCL_DEVICE", with value "off" (no quotes). I expect Windows 10 is similar.

**Reply by dlemstra** [reply]
The latest release of ImageMagick should disable OpenCL by default. Will need to investigate why this is failing.

**Reply by flyerx** [reply]
That explains some functions not working and some working fine.  I though I was going crazy. 

I just updated the env variable to make it permanent. 

Just in case someone is using powershell and finds this thread useful the environment variable can be set like this:
 $env:MAGICK_OCL_DEVICE='off'

the `set MAGICK_OCL_DEVICE=off` works in the old command prompt but not powershell. 

Thanks for the help.

**Reply by dlemstra** [reply]
It appears that I made a mistake when disabling OpenCL by default. This will be resolved in the next release of ImageMagick.

**Reply by ghost** [reply]
> That explains some functions not working and some working fine. I though I was going crazy.

You're not alone @flyerx. I too thought I was going crazy because I would see the command run, not see an error but not get an output file either. What made the feeling like I was doing something wrong worse was when certain commands would work and a simple resize didn't that I actually began to question my ability to use a CLI. 😅
